### ISSUE TITLE
Handling JSON.parse exceptions in LokiPartitioningAdapter

### ISSUE THREAD

**Issue by dnalor** [techfort/LokiJS#552]
If a non-existing database is loaded:

```
LokiPartitioningAdapter.prototype.loadDatabase 
//var db = JSON.parse(result);
        var db = null;
		try {
			db = JSON.parse(result);
		} catch(e) {
          callback(new Error("LokiPartitioningAdapter loadDatabase() no results"));
		  return;
		}

```
If there are problems with the JSON data (in my case (NativeScript app), for unknown reasons, there seem to be mis-encoded chars in the file, even though it was created by saveDatabase):
```
LokiPartitioningAdapter.prototype.loadNextPage 
for(idx=0; idx < dlen; idx++) {
	  try{
          self.dbref.collections[self.pageIterator.collection].data.push(JSON.parse(data[idx]));
	  } catch(e) {
		  console.log("LokiPartitioningAdapter.prototype.loadDatabase JSON.parse exception="+e);
		  console.log("LokiPartitioningAdapter.prototype.loadDatabase JSON data="+data[idx]);
	  }
          data[idx] = null;
        }
```

**Comment by obeliskos** [comment]
Are you able to inspect the contents of the persisted files to see why they are seem to be 'mis-encoded' chars?

If there is a clash with the default delimeter '$>' that can be overridden using the 'destructureDelimiter' option of the [loki constructor](https://rawgit.com/techfort/LokiJS/master/jsdoc/Loki.html) options.

**Comment by dnalor** [comment]
After further investigation, it seems that there is simply always a parse exception for the last line of a partition, unrelated to it's content:
```
[cut]
JS: LokiPartitioningAdapter.prototype.loadNextPage keyname=/data/user/0/net.nexus0.lms.LMSlib2go/files/lms.json.2.12
JS: LokiPartitioningAdapter.prototype.loadDatabase result=3632258 delimiter=$<
JS: LokiPartitioningAdapter.prototype.loadNextPage after split dlen=4202
JS: LokiPartitioningAdapter.prototype.loadNextPage isLastPage=false
JS: LokiPartitioningAdapter.prototype.loadNextPage JSON.parse exception=SyntaxError: Unexpected token $ in JSON at position 795
JS: LokiPartitioningAdapter.prototype.loadNextPage JSON data={"trackId":427172,"trackTitle":"éä½","trackArtist":"Kibaco","trackCoverid":"d08c644d","trackCompilation":"0","trackDuration":247.1,"trackAlbumId":"39024","trackFilesize":"9886361","trackGenre":"Electronic","trackGenres":"Electronic","trackDisc":null,"trackArtworkTrackId":"d08c644d","trackComment":null,"trackAlbum":"Heiwa No Uta (INC)","trackModificationTime":"1237268732","trackGenreId":"754","trackGenreIds":"754","trackAddedTime":"1490830055","trackDisccount":null,"trackBitrate":"320kbps CBR","trackRating":null,"trackPlaycount":null,"trackArtistId":"26321","trackTracknum":"4","trackSamplerate":"44100","trackUrl":"file:///data/music/Kibaco%20-%20Heiwa%20No%20Uta%20(INC)/04.mp3","trackLyrics":null,"trackYear":"2007","meta":{"revision":0,"created":1491329620101,"version":0},"$loki":77456}$<
JS: LokiPartitioningAdapter.prototype.loadNextPage data pushed
JS: LokiPartitioningAdapter.prototype.loadNextPage keyname=/data/user/0/net.nexus0.lms.LMSlib2go/files/lms.json.2.13
JS: LokiPartitioningAdapter.prototype.loadDatabase result=0 delimiter=$<
JS: LokiPartitioningAdapter.prototype.loadNextPage after split dlen=1
JS: LokiPartitioningAdapter.prototype.loadNextPage isLastPage true
```
The last file (13) is actually empty, and seems to have been created when loading the DB, and not when is was saved (notice the time difference):
```
-rw------- 1 98 98 5244659 Apr  4 20:31 lms.json.2.0
-rw------- 1 98 98 5244478 Apr  4 20:31 lms.json.2.1
-rw------- 1 98 98 5244161 Apr  4 20:31 lms.json.2.10
-rw------- 1 98 98 5244235 Apr  4 20:31 lms.json.2.11
-rw------- 1 98 98 3633164 Apr  4 20:31 lms.json.2.12
-rw------- 1 98 98       0 Apr  4 21:19 lms.json.2.13
-rw------- 1 98 98 5244132 Apr  4 20:31 lms.json.2.2
-rw------- 1 98 98 5245116 Apr  4 20:31 lms.json.2.3
-rw------- 1 98 98 5245006 Apr  4 20:31 lms.json.2.4
-rw------- 1 98 98 5243773 Apr  4 20:31 lms.json.2.5
-rw------- 1 98 98 5244089 Apr  4 20:31 lms.json.2.6
-rw------- 1 98 98 5244112 Apr  4 20:31 lms.json.2.7
-rw------- 1 98 98 5243589 Apr  4 20:31 lms.json.2.8
-rw------- 1 98 98 5243732 Apr  4 20:31 lms.json.2.9
```
This happens for all collections.
I had a look at the destructureDelimiter, but it seems that the LokiPartitioningAdapter uses it's own delimiter. Setting both to something else seems to make no difference.

**Comment by obeliskos** [comment]
Is that a 5 meg page size?  Is that a nativescript limit?

Yea for the LokiPartitioningAdapter, I guess you would use the delimeter option off of the [LokiPartitioningAdapter constructor](https://rawgit.com/techfort/LokiJS/master/jsdoc/LokiPartitioningAdapter.html) instead.

Is there any way to reproduce this using partitioning + memory adapter on a desktop or sharable sandbox gist or link?

Does your app immediately exit after saving the database?

**Comment by dnalor** [comment]
> Is that a 5 meg page size? Is that a nativescript limit?

Yes it's 5MB. I get a OutOfMemory exception when saving larger chunks (that was the reason I switched to the LokiPartitioningAdapter in the first place). 
Probably not a nativescript limit - the app is running in an Android emulator with enough memory, but the VM seems to be unable to keep up with a rapid succession of allocation requests (not specific to Loki - I have the same issue when fetching data from the server).

> Is there any way to reproduce this using partitioning + memory adapter on a desktop or sharable sandbox gist or link?

While I'm developing this on a desktop, I'm not quite sure how to accomplish this in a simple way. Basically, you'd need

- the app running in the Android emulator 
- the nativescript install (for debugging)
- the server side (which is a Logitech Media Server)

While all of this is freely available, depending on your current setup, this might require some effort.
 
> Does your app immediately exit after saving the database?

Yes it does, there is only a single saveDatabase call when it is closed.
The zero byte file is created after starting the app again, at which point it calls loadDatabase (also only once).

**Comment by dnalor** [comment]
After some more testing, it seems the empty files are created by the NativeScript function used to read files (which creates a new file when trying to read a non-existing file).
So it seems that the LokiPartitioningAdapter always tries to read one more file (page) than it should.
The cause for this is that, since it always fails on the last entry, it also fails to properly detect the last page (which only contains one line)
Here's the flow with a simple collection containing only this document:
```
appVersion":1,"confKey":1,"appTheme":"dark","apiUrl":"http://raspi3-64:9000/jsonrpc.js","maxFetchItems":5601,"fetchChunks":2800,"warnIfNoWifi":false,"plsRegex":"file:///data/music","plsSubst":"/storage/3661-6535/music","plsPath":"/storage/8024-0800/data","meta":{"revision":0,"created":1492808911993,"version":0},"$loki":1}$<
```
Output:
```
JS: LokiPartitioningAdapter.prototype.loadNextPartition keyname=/data/user/0/net.nexus0.lms.LMSlib2go/files/lms.json.3 paging=true
JS: LokiPartitioningAdapter.prototype.loadNextPage keyname=/data/user/0/net.nexus0.lms.LMSlib2go/files/lms.json.3.0
JS: LokiNativescriptAdapter.loadDatabase dbname=/data/user/0/net.nexus0.lms.LMSlib2go/files/lms.json.3.0 callback=
JS: LokiPartitioningAdapter.loadNextPage result.length=327 delimiter=$<
// result from LokiNativescriptAdapter
JS: LokiPartitioningAdapter.loadNextPage result={"appVersion":1,"confKey":1,"appTheme":"dark","apiUrl":"http://raspi3-64:9000/jsonrpc.js","maxFetchItems":5601,"fetchChunks":2800,"warnIfNoWifi":false,"plsRegex":"file:///data/music","plsSubst":"/storage/3661-6535/music","plsPath":"/storage/8024-0800/data","meta":{"revision":0,"created":1492808911993,"version":0},"$loki":1}$<
// after .split()
JS: LokiPartitioningAdapter.loadNextPage data[0]={"appVersion":1,"confKey":1,"appTheme":"dark","apiUrl":"http://raspi3-64:9000/jsonrpc.js","maxFetchItems":5601,"fetchChunks":2800,"warnIfNoWifi":false,"plsRegex":"file:///data/music","plsSubst":"/storage/3661-6535/music","plsPath":"/storage/8024-0800/data","meta":{"revision":0,"created":1492808911993,"version":0},"$loki":1}$<
JS: LokiPartitioningAdapter.loadNextPage data[dlen-1]={"appVersion":1,"confKey":1,"appTheme":"dark","apiUrl":"http://raspi3-64:9000/jsonrpc.js","maxFetchItems":5601,"fetchChunks":2800,"warnIfNoWifi":false,"plsRegex":"file:///data/music","plsSubst":"/storage/3661-6535/music","plsPath":"/storage/8024-0800/data","meta":{"revision":0,"created":1492808911993,"version":0},"$loki":1}$<
JS: LokiPartitioningAdapter.prototype.loadNextPage isLastPage false
JS: LokiPartitioningAdapter.prototype.loadNextPage JSON.parse exception=SyntaxError: Unexpected token $ in JSON at position 325
JS: LokiPartitioningAdapter.prototype.loadNextPage JSON data={"appVersion":1,"confKey":1,"appTheme":"dark","apiUrl":"http://raspi3-64:9000/jsonrpc.js","maxFetchItems":5601,"fetchChunks":2800,"warnIfNoWifi":false,"plsRegex":"file:///data/music","plsSubst":"/storage/3661-6535/music","plsPath":"/storage/8024-0800/data","meta":{"revision":0,"created":1492808911993,"version":0},"$loki":1}$<
JS: LokiPartitioningAdapter.prototype.loadNextPage data pushed
JS: LokiPartitioningAdapter.prototype.loadNextPage keyname=/data/user/0/net.nexus0.lms.LMSlib2go/files/lms.json.3.1
JS: LokiNativescriptAdapter.loadDatabase dbname=/data/user/0/net.nexus0.lms.LMSlib2go/files/lms.json.3.1 callback=
JS: LokiPartitioningAdapter.loadNextPage result.length=2 delimiter=$<
JS: LokiPartitioningAdapter.loadNextPage result=$<
JS: LokiPartitioningAdapter.loadNextPage data[0]=$<
// isLastPage will be incorrect
JS: LokiPartitioningAdapter.loadNextPage data[dlen-1]=$<
JS: LokiPartitioningAdapter.prototype.loadNextPage isLastPage false
JS: LokiPartitioningAdapter.prototype.loadNextPage JSON.parse exception=SyntaxError: Unexpected token $ in JSON at position 0
JS: LokiPartitioningAdapter.prototype.loadNextPage JSON data=$<
JS: LokiPartitioningAdapter.prototype.loadNextPage data pushed
// attempt to read non-existing next page 
JS: LokiPartitioningAdapter.prototype.loadNextPage keyname=/data/user/0/net.nexus0.lms.LMSlib2go/files/lms.json.3.2
JS: LokiNativescriptAdapter.loadDatabase dbname=/data/user/0/net.nexus0.lms.LMSlib2go/files/lms.json.3.2 callback=
JS: LokiPartitioningAdapter.loadNextPage result.length=0 delimiter=$<
JS: LokiPartitioningAdapter.loadNextPage result=
JS: LokiPartitioningAdapter.loadNextPage data[0]=
JS: LokiPartitioningAdapter.loadNextPage data[dlen-1]=
JS: LokiPartitioningAdapter.prototype.loadNextPage isLastPage true
JS: LokiPartitioningAdapter.prototype.loadNextPage data pushed
```
It seems that the split fails (data[0] contains the delimiter). The expected result would be:
```
'{"appVersion":1,"confKey":1,"appTheme":"dark","apiUrl":"http://raspi3-64:9000/jsonrpc.js","maxFetchItems":5601,"fetchChunks":2800,"warnIfNoWifi":false,"plsRegex":"file:///data/music","plsSubst":"/storage/3661-6535/music","plsPath":"/storage/8024-0800/data","meta":{"revision":0,"created":1492808911993,"version":0},"$loki":1}$<'.split('$<');

0:"{"appVersion":1,"confKey":1,"appTheme":"dark","apiUrl":"http://raspi3-64:9000/jsonrpc.js","maxFetchItems":5601,"fetchChunks":2800,"warnIfNoWifi":false,"plsRegex":"file:///data/music","plsSubst":"/storage/3661-6535/music","plsPath":"/storage/8024-0800/data","meta":{"revision":0,"created":1492808911993,"version":0},"$loki":1}"
1:""
length:2
```

**Comment by obeliskos** [comment]
Hmm my initial thoughts are that it could be either : 
- our delimeter contains a newline and that is being handled differently in your environment
- your storage adapter is asynchronous and you are ending the task before it can complete  (maybe wait a few seconds after save before closing app as test?)

The other option is that we have a bug somewhere... or we need to handle something better.

I did revise a partitioning example gist which you can [view here](https://rawgit.com/techfort/LokiJS/master/examples/sandbox/LokiSandbox.htm#rawgist=https://gist.githubusercontent.com/obeliskos/ff2687e8496ec53a8a74ba47bb4eefa0/raw/4e6b2ac8ce7940b1f7e2b8e6a63d6efd4a61751e/PagePartitionedMemory).

That example persists to an in memory adapter which is synchronous but obviously volatile... maybe you can try similar logic and inspect its contents to see if the partitions get structured differently for that single test case.

So if we can isolate and rule out asynchronous storage issues (maybe add save button and wait until it has a chance to complete before exiting) and rule out newline differences (could try a delimiter without newline escape sequence which i put in just for easy reading) then we can see if your data presents a unique case which exploits a real bug.

**Comment by dnalor** [comment]
> your storage adapter is asynchronous and you are ending the task before it can complete

I don't think so
- I have callbacks for saveDatabase() and close(), both are executed and no error is passed to them
- I actually already added a save button for testing purposes, so I can be sure the app continues running after the save
- I have mounted the emulator's file system, so I can see that the files are not truncated

The document posted above is directly copied from the file and seems to be complete.
Here's the hex dump of the on-disc file:
```
00000000 7B 22 61 70 | 70 56 65 72 | 73 69 6F 6E | 22 3A 31 2C | 22 63 6F 6E | 66 4B 65 79 | 22 3A 31 2C | 22 61 70 70  {"appVersion":1,"confKey":1,"app
00000020 54 68 65 6D | 65 22 3A 22 | 64 61 72 6B | 22 2C 22 61 | 70 69 55 72 | 6C 22 3A 22 | 68 74 74 70 | 3A 2F 2F 72  Theme":"dark","apiUrl":"http://r
00000040 61 73 70 69 | 33 2D 36 34 | 3A 39 30 30 | 30 2F 6A 73 | 6F 6E 72 70 | 63 2E 6A 73 | 22 2C 22 6D | 61 78 46 65  aspi3-64:9000/jsonrpc.js","maxFe
00000060 74 63 68 49 | 74 65 6D 73 | 22 3A 35 36 | 30 31 2C 22 | 66 65 74 63 | 68 43 68 75 | 6E 6B 73 22 | 3A 32 38 30  tchItems":5601,"fetchChunks":280
00000080 30 2C 22 77 | 61 72 6E 49 | 66 4E 6F 57 | 69 66 69 22 | 3A 66 61 6C | 73 65 2C 22 | 70 6C 73 52 | 65 67 65 78  0,"warnIfNoWifi":false,"plsRegex
000000A0 22 3A 22 66 | 69 6C 65 3A | 2F 2F 2F 64 | 61 74 61 2F | 6D 75 73 69 | 63 22 2C 22 | 70 6C 73 53 | 75 62 73 74  ":"file:///data/music","plsSubst
000000C0 22 3A 22 2F | 73 74 6F 72 | 61 67 65 2F | 33 36 36 31 | 2D 36 35 33 | 35 2F 6D 75 | 73 69 63 22 | 2C 22 70 6C  ":"/storage/3661-6535/music","pl
000000E0 73 50 61 74 | 68 22 3A 22 | 2F 73 74 6F | 72 61 67 65 | 2F 38 30 32 | 34 2D 30 38 | 30 30 2F 64 | 61 74 61 22  sPath":"/storage/8024-0800/data"
00000100 2C 22 6D 65 | 74 61 22 3A | 7B 22 72 65 | 76 69 73 69 | 6F 6E 22 3A | 30 2C 22 63 | 72 65 61 74 | 65 64 22 3A  ,"meta":{"revision":0,"created":
00000120 31 34 39 32 | 39 37 38 34 | 36 38 39 38 | 30 2C 22 76 | 65 72 73 69 | 6F 6E 22 3A | 30 7D 2C 22 | 24 6C 6F 6B  1492978468980,"version":0},"$lok
00000140 69 22 3A 31 | 7D 24 3C 0A |                                                                                    i":1}$<.
```
I can't see anything wrong with it - delimiter present, *nix EOL correct.
What really baffles me is that it's only the last line - I have pages with eg. 7000 lines, 6999 are processed correctly (i.e. the .split() works on them)

If you look at my previous comment:
```
JS: LokiPartitioningAdapter.loadNextPage result={"appVersion":1,"confKey":1,"appTheme":"dark","apiUrl":"http://raspi3-64:9000/jsonrpc.js","maxFetchItems":5601,"fetchChunks":2800,"warnIfNoWifi":false,"plsRegex":"file:///data/music","plsSubst":"/storage/3661-6535/music","plsPath":"/storage/8024-0800/data","meta":{"revision":0,"created":1492808911993,"version":0},"$loki":1}$<
// after .split()
JS: LokiPartitioningAdapter.loadNextPage data[0]={"appVersion":1,"confKey":1,"appTheme":"dark","apiUrl":"http://raspi3-64:9000/jsonrpc.js","maxFetchItems":5601,"fetchChunks":2800,"warnIfNoWifi":false,"plsRegex":"file:///data/music","plsSubst":"/storage/3661-6535/music","plsPath":"/storage/8024-0800/data","meta":{"revision":0,"created":1492808911993,"version":0},"$loki":1}$<
```
the adapter reads the file correct as far as I can see, but the split() fails. Even if there would be a special character in the file which affects the split but cannot be seen in the console output, the hex dump should have shown it.

**Comment by pablocid** [comment]
change the delimiter option in LokiPartitioningAdapter
```
new LokiPartitioningAdapter(new LokiNativescriptAdapter(), { paging: true, pageSize: 5 * 1024 * 1024, delimiter:'$<-|->$' }), 
```

**Comment by dnalor** [comment]
Very interesting - this seems to work. The difference to the delimiters I tried seems to be that this one doesn't contain an '\n'.
Do you know if this is the reason why it works?

**Comment by pablocid** [comment]
Yes, is the reason , I think it is a parse problem. 
I had another problem with Loki constructor . In LokiPartitioningAdapter definition (lokijs.js), the dbref is created with new Loki(dbname), and gave me the adapter error. I revolve that modifying de constructor:  new Loki(dbname, {})

**Comment by pablocid** [comment]
Sorry, I think the parsing problem is when writing the file and don't retain the new line. I don't know if it is java thing (Android) or Javascript thing. Look this: [unable-to-retain-line-breaks-when-writing-text-file-as-blob](https://stackoverflow.com/questions/19190331/unable-to-retain-line-breaks-when-writing-text-file-as-blob)

**Comment by dnalor** [comment]
> I had another problem with Loki constructor . In LokiPartitioningAdapter definition (lokijs.js), the dbref is created with new Loki(dbname), and gave me the adapter error. I revolve that modifying de constructor: new Loki(dbname, {})

Could be related to https://github.com/techfort/LokiJS/issues/553

**Comment by dnalor** [comment]
> Sorry, I think the parsing problem is when writing the file and don't retain the new line. I don't know if it is java thing (Android) or Javascript thing. Look this: unable-to-retain-line-breaks-when-writing-text-file-as-blob

The stackoverflow issue is about downloading text as a file in a browser, so a very different case.
As mentioned in https://github.com/techfort/LokiJS/issues/552#issuecomment-296489724, the new line characters seem to be written correctly to the file, reading the file back fails with a parse exception on the last line

**Comment by obeliskos** [comment]
@pablocid : my trace though loki constructor would 'appear' to handle not passing that options parameter but we should probably force it early on (in loki constructor) to be : 
```javascript
options = options || {};
```
If that fixes it for you, we can patch that in sometime.

@dnalor : what is the status of your original issue, is it fixed by not using newline characters?  Or does it still try to read an extra file (thus creating it?)

**Comment by dnalor** [comment]
@obeliskos :
> what is the status of your original issue, is it fixed by not using newline characters? Or does it still try to read an extra file (thus creating it?) 

Without a newline, it seems to work flawlessly (tested both on an Android emulator and an actual phone).
I'd rather call it a workaround than a fix, though - while I'm happy it works for me now, for anyone who is unaware of this, it's still broken.
(I closed the issue by accident - hit the wrong button :-/)

>If so we may need you to add diagnostic trace/console statements in the LokiPartitioningAdapter.prototype.loadNextPage function. This function splits the data, removes any empty last element, and determines if we have another page which needs to be loaded.

I tried that, see https://github.com/techfort/LokiJS/issues/552#issuecomment-296362071
Not sure if the output is easy to understand, I added some comments for clarity.
Basically, I think everything after the failed split are subsequent errors which can be disregarded since they wouldn't occur otherwise i.e.

loadNextPage (on last page)->split (fails on last line)->JSON parse exception, data from line lost->last page not detected->attempt to read non-existing next page->NativeScript adapter creates empty file ([NativeScript File.readText](http://docs.nativescript.org/api-reference/classes/_file_system_.file.html#readtext) behaviour)->empty file read->no data,  last page detected

It boils down to why the split fails on the last line when using a newline as a delimiter.
